File: Leetcode Daily Challenge/026-August-2025/011-range-product-queries-of-powers.java
// https://leetcode.com/problems/range-product-queries-of-powers/




import java.util.*;

class Solution {
    private final int M = 1_000_000_000 + 7;

    // TC : O(Q)
    // SC : O(1)
    public int[] productQueries(int n, int[][] queries) {
        List<Integer> powers = new ArrayList<>();
        for (int i = 0; i <= 31; i ++) {
            if ((n & (1 << i)) != 0) {
                int val = 1 << i; // (int) Math.pow(2, i);
                powers.add(val);
            }
        }    
        
        int q = queries.length;
        int[] result = new int[q];
        for (int i = 0; i < q; i ++) {
            int l = queries[i][0], r = queries[i][1];
            long val = 1;
            for (int j = l; j <= r; j ++) {
                val = (val * powers.get(j)) % M;
            }
            result[i] = (int) val;
        }
        return result;
    }
}

File: Leetcode Daily Challenge/026-August-2025/011-range-product-queries-of-powers.cpp
// https://leetcode.com/problems/range-product-queries-of-powers/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    const int M = 1e9 + 7;
    
    // TC : O(q)
    // SC : O(1)
    vector<int> productQueries(int n, vector<vector<int>>& queries) {
        vector<int> powers;
        for (int i = 0; i <= 31; i ++) {
            if (n & (1 << i)) {
                powers.push_back(1 << i);
            }
        }

        vector<int> result;
        for (auto &q : queries) {
            int l = q[0], r = q[1];
            long long val = 1;
            for (int i = l; i <= r; i ++) {
                val = (val * powers[i]) % M;
            }
            result.push_back(val);
        }
        return result;
    }
};